src/generate_features.py

```diff
@@ -1,3 +1,215 @@
+    top_n = config.get('popular_threshold', 5)
+
+    top_products = products.groupby('department_id')['n_orders']\
+                   .nlargest(top_n).reset_index()
+    top_products = top_products.join(products[['product_name']], on='product_id')
+    top_products = top_products.values.reshape(-1)
+    products['popular'] = products['product_name'].isin(top_products).astype(int)
+
+    products['organic'] = products['product_name'].str.lower()\
+                          .str.contains('organic').astype(int)
+
+    # cats contains mapping of macro-level category (vegetable, beverage, etc) to
+    # all aisles which could fall under that classification
+    for cat in cats:
+        aisles[cat] = aisles['aisle'].isin(cat_map[cat]).astype(int)
+
+    logger.debug('Engineering order features')
+    orders = orders.join(order_products['order_id'].value_counts().rename('order_size'))
+
+    # discretize order size to help with long-tail issue
+    # these categories not currently used
+    orders['size_cat'] = pd.cut(orders['order_size'],
+                                [0, 5, 10, 20, np.inf],
+                                labels=['small', 'medium', 'large', 'xl'])
+
+    full = order_products.join(orders, on='order_id')
+    full = full.join(products, on='product_id')
+    full = full.join(aisles, on='aisle_id')
+
+    groups = full.groupby('order_id')
+
+    # order_types will be combination of metadata and product-level features
+    order_types = groups[cats + ['reordered', 'organic', 'popular']].mean()
+    meta_cols = ['order_dow', 'order_hour_of_day', 'days_since_prior_order',
+                 'eval_set', 'order_number', 'order_size', 'user_id']
+    order_types = order_types.join(orders[meta_cols])
+    # strip away metadata and categoricals for clustering
+    X = order_types.drop(columns=['order_dow', 'order_hour_of_day',
+                                  'days_since_prior_order', 'eval_set',
+                                  'order_number', 'user_id']).values
+
+    if config.get('scale_x'):
+        logger.debug('Scaling order features')
+        X = preprocessing.scale(X)
+
+    # filter out rows with NaN vals
+    X = clean(X)
+    logger.debug('Clustering orders based on %d orders and %d variables',
+                 X.shape[0], X.shape[1])
+
+    if config.get('cluster-method') == 'gmm':
+        n_components = config.get('n_components')
+        cov_type = config.get('cov_type')
+        labels = gmm_clust(X, n_components, cov_type)
+    elif config.get('cluster-method') == 'kmeans':
+        k = config.get('k')
+        labels = km_clust(X, k)
+
+    logger.debug('Cluster solution score: %f', score_ch(X, labels))
+    logger.info('Cluster dispersion: %s',
+                np.unique(labels, return_counts=True))
+
+    order_types['label'] = labels
+
+    if config.get('save_cluster_heatmap'):
+        gen_plots(order_types)
+
+    # Limit orders to "prior" set for training data
```

Comment:
whats are sizes of test/train

Reply:
"prior" is a tag in the raw data that is given to every order in the dataset except for each user's most recent order - if this is not documented anywhere I will be sure to clarify